vital/algo/serialize_metadata.cxx

```diff
@@ -0,0 +1,114 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/**
+ * \file
+ * \brief Implementation of load/save wrapping functionality.
+ */
+
+#include "serialize_metadata.h"
+
+#include <vital/algo/algorithm.txx>
+#include <vital/exceptions/io.h>
+#include <vital/vital_config.h>
+#include <vital/vital_types.h>
+
+#include <kwiversys/SystemTools.hxx>
+
+/// \cond DoxygenSuppress
+INSTANTIATE_ALGORITHM_DEF( kwiver::vital::algo::serialize_metadata );
+/// \endcond
+
+namespace kwiver {
+
+namespace vital {
+
+namespace algo {
+
+serialize_metadata
+::serialize_metadata()
+{
+  attach_logger( "algo.serialize_metadata" );
+}
+
+metadata_map_sptr
+serialize_metadata
+::load( std::string const& filename ) const
+{
+  // Make sure that the given file path exists and is a file.
+  if( !kwiversys::SystemTools::FileExists( filename ) )
+  {
+    VITAL_THROW( path_not_exists,
+                 filename );
+  }
+  else if( kwiversys::SystemTools::FileIsDirectory( filename ) )
+  {
+    VITAL_THROW( path_not_a_file,
+                 filename );
+  }
+
+  return this->load_( filename );
+}
+
+void
+serialize_metadata
+::save( std::string const& filename,
+        metadata_map_sptr data ) const
+{
+  // Make sure that the given file path's containing directory exists and is
+  // actually a directory.
+  std::string containing_dir =
+    kwiversys::SystemTools::GetFilenamePath(
+      kwiversys::SystemTools::CollapseFullPath( filename ) );
+
+  if( !kwiversys::SystemTools::FileExists( containing_dir ) )
+  {
+    VITAL_THROW( path_not_exists,
+                 containing_dir );
+  }
+  else if( !kwiversys::SystemTools::FileIsDirectory( containing_dir ) )
+  {
+    VITAL_THROW( path_not_a_directory,
+                 containing_dir );
+  }
+
+  this->save_( filename,
+               data );
+}
+
+const vital::algorithm_capabilities&
```

Comment:
```suggestion
vital::algorithm_capabilities
```

Note the deliberate removal of `const&` (will need to change in header also).

Reply:
Is this a danger for const-qualified methods?

Reply:
If you return a reference, the caller might accept the value as a reference. If the owning object then goes away, you have a problem. We've had issues with this in the past.

I ended up not changing this because, although this isn't actually a `virtual` method, the same API does exist elsewhere, so for the moment I elected to keep it consistent. We should, however, consider changing it, but ideally we would change all of them, so we may want to revisit this as a follow-up.